docs: add missing standards for building components (#22412)
angular/components: CODING_STANDARDS.md

}
```

##### Selectors
* Component selectors should be lowercase and delimited by hyphens.
* Directive selectors should be camel cased.

#### Inheritance

Avoid using inheritance to apply reusable behaviors to multiple components. This limits how many
behaviors can be composed. Instead, [TypeScript mixins][ts-mixins] can be used to compose multiple
common behaviors into a single component.

#### MDC checks
To ensure backwards compatability, we check that tests written for MDC Components include all of
the same tests that the non-MDC version had. Similarly, we check that the public API of MDC
Components match that of the non-MDC version.

In the case where old tests no longer make sense and should be omitted or the public API should be
changed, you can do so in [scripts/check-mdc-tests-config.ts](https://github.com/angular/components/blob/master/scripts/check-mdc-tests-config.ts) and
[scripts/check-mdc-exports-config.ts](https://github.com/angular/components/blob/master/scripts/check-mdc-exports-config.ts).
Remember to leave a comment explaining why the change was necessary.

#### Coercion

> When defining public properties, use our coercion API to coerce `@Input`s into specific types.

Reviewer: I think that this is covered by the `Boolean arguments` section already.

Author: From what I can tell, the `Boolean arguments` section only talks about how we should avoiding adding boolean arguments to a method in cases where the argument means "do something extra". Can you provide a link to where coercion is already documented?

Reviewer: You're right, I didn't read it through. One note on your change is it makes it sound like we should be coercing all public properties, but it's only relevant for numbers and booleans.

For example:
```ts
@Input() disabled: boolean;
get disabled(): boolean { return this._disabled; }
set disabled(v: boolean) { this._disabled = coerceBooleanProperty(v); }
private _disabled = false;

...

static ngAcceptInputType_value: BooleanInput;
```
The above code would allow users to set `disabled` similar to how it can be set on native inputs:
```html
<component disabled></component>
```
Even though an empty string is technically what is being provided as the value of `disabled`,
`ngAcceptInputType` allows the mismatched type to be provided and `coerceBooleanProperty`
interprets the given value (an empty string) to the correct type & value, which in this case would
be `true`.

> #### Expose native inputs

Reviewer: I'm not sure that somebody new to the project would know what this refers to. Maybe add a `Do` and `Don't` example?

}
```

##### Selectors

> * Component selectors should be lowercase and delimited by hyphens.

Reviewer 2: I would expand this to say:
"Components should use element selectors except when the component API uses a native HTML element."

> * Directive selectors should be camel cased.

Reviewer 2: I would add "Exceptions may be made for directives that _act_ like a component but would have an empty template, or when the directive is intended to match some existing attribute."

#### Inheritance

Avoid using inheritance to apply reusable behaviors to multiple components. This limits how many
behaviors can be composed. Instead, [TypeScript mixins][ts-mixins] can be used to compose multiple
common behaviors into a single component.

#### MDC checks
To ensure backwards compatability, we check that tests written for MDC Components include all of
the same tests that the non-MDC version had. Similarly, we check that the public API of MDC
Components match that of the non-MDC version.

In the case where old tests no longer make sense and should be omitted or the public API should be
changed, you can do so in [scripts/check-mdc-tests-config.ts](https://github.com/angular/components/blob/master/scripts/check-mdc-tests-config.ts) and
[scripts/check-mdc-exports-config.ts](https://github.com/angular/components/blob/master/scripts/check-mdc-exports-config.ts).
Remember to leave a comment explaining why the change was necessary.

#### Coercion

> When defining public properties that are booleans or numbers, use our coercion API to coerce

Reviewer 2: ```suggestion
Component and directive inputs for boolean and number values must use a setter to
coerce values to the expected type using cdk/coercion.
```

`@Input`s into specific types.
For example:
```ts
@Input() disabled: boolean;
get disabled(): boolean { return this._disabled; }
set disabled(v: boolean) { this._disabled = coerceBooleanProperty(v); }
private _disabled = false;

...

static ngAcceptInputType_value: BooleanInput;
```

> The above code would allow users to set `disabled` similar to how it can be set on native inputs:

Reviewer 2: ```suggestion
The above code allows users to set `disabled` similar to how it can be set on native inputs:
```

```html
<component disabled></component>
```
Even though an empty string is technically what is being provided as the value of `disabled`,
`ngAcceptInputType` allows the mismatched type to be provided and `coerceBooleanProperty`
interprets the given value (an empty string) to the correct type & value, which in this case would
be `true`.

#### Expose native inputs
Native inputs used in components should be exposed to developers through `ng-content`. This allows
developers to interact directly with the input and allows us to avoid providing custom
implementations for all of the input's native behaviors.

For example:

**Do:**

Implementation
```html
<ng-content></ng-content>
```

Usage
```html

> <yourComponent>

Reviewer 2: `your-component`

```

[ts-mixins]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-mix-in-classes

#### Prefer CSS classes to tag names for styling

> TODO(wagnermaciel): Ask @jelbourn why we prefer this.

Reviewer: One of the explanations could be that targeting the tag name will cause a conflict with the MDC version of the component.

```

[ts-mixins]: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-2.html#support-for-mix-in-classes

> #### Prefer CSS classes to tag names for styling

Reviewer 2: ```suggestion
#### Prefer CSS classes to tag names and attributes for styling
```

Targeting tag names can cause conflicts with the MDC version of the component. For this reason, use

> css class names defined by us instead of tag names.

Reviewer 2: ```suggestion
CSS class names defined by us instead of tag names.
```

```scss
/** Do: */
.mat-mdc-slider { ... }

/** Don't: */
mdc-slider { ... }

> ```

Reviewer 2: We also prefer classes over attributes for consistency

Author: @jelbourn Do you want me to add exactly this to the description?

```
Targeting tag names can cause conflicts with the MDC version of the component. For this reason, use
CSS class names defined by us instead of tag names. We also prefer classes over attributes for
consistency.
```

Reviewer 2: Sure, that works
